Source/WebCore/html/canvas/PaintRenderingContext2D.cpp

```diff
@@ -26,22 +26,64 @@
 #include "config.h"
 #include "PaintRenderingContext2D.h"
 
+#include "CustomPaintCanvas.h"
+#include "DisplayListDrawingContext.h"
+#include "DisplayListRecorder.h"
+#include "DisplayListReplayer.h"
 #include <wtf/IsoMallocInlines.h>
 
 namespace WebCore {
 
 WTF_MAKE_ISO_ALLOCATED_IMPL(PaintRenderingContext2D);
 
-std::unique_ptr<PaintRenderingContext2D> PaintRenderingContext2D::create(CanvasBase& canvas)
+std::unique_ptr<PaintRenderingContext2D> PaintRenderingContext2D::create(CustomPaintCanvas& canvas)
 {
     return std::unique_ptr<PaintRenderingContext2D>(new PaintRenderingContext2D(canvas));
 }
 
-PaintRenderingContext2D::PaintRenderingContext2D(CanvasBase& canvas)
+PaintRenderingContext2D::PaintRenderingContext2D(CustomPaintCanvas& canvas)
     : CanvasRenderingContext2DBase(canvas, { }, false)
 {
 }
 
 PaintRenderingContext2D::~PaintRenderingContext2D() = default;
 
+CustomPaintCanvas& PaintRenderingContext2D::canvas() const
+{
+    return downcast<CustomPaintCanvas>(canvasBase());
```

Comment:
So `PaintRenderingContext2D` will never use the ImageBuffer which is owned by CanvasBase which is the base class of `CustomPaintCanvas`. right?

Reply:
Correct.